Further enhancements to the Bech32 library.

jonathanknowles opened the pull request:
## Issue Number

Issue #238 

## Overview

**This PR:**

1. Implements **error location detection** for the Bech32 encoding, based on the JavaScript [reference implementation](https://github.com/sipa/bech32/blob/master/ecc/javascript/bech32_ecc.js).

2. Provides a set of **property tests** to verify that corrupted Bech32 strings are correctly **rejected** by the `decode` function. Corruption is simulated in various ways:
    * deleting characters
    * inserting characters
    * swapping characters
    * mutating characters

## Caveats

1. The error location detection is **best effort**:
    * In cases where it **is** possible to unambiguously identify the locations of erroneous characters, the decoder will return the `StringToDecodeContainsInvalidChars` error with a suitable list of locations.
    * In cases where it's **not** possible to unambiguously identify the locations of erroneous characters, the decoder will return the `StringToDecodeContainsInvalidChars` with the empty list.

2. Currently, there is no automated test to confirm that the `locateErrors` and `syndrome` functions behave identically to the equivalent functions in the reference JavaScript implementation. :warning:  **It is possible that these implementations differ** (and this is something that we should pay attention to).

Nevertheless, in the test suite, we verify that **if** the position of an invalid character **is** detected, then it is **detected at the correct location**.

KtorZ commented on src/Cardano/Wallet/Kernel/DB/AcidState.hs:
```diff
@@ -502,69 +499,29 @@ observableRollbackUseInTestsOnly = runUpdateDiscardSnapshot $
-               -- 'HdAccount'.
-               -> Maybe HdAddress
-               -- ^ An optional default HdAddress to go with this HdRoot.
-               -> Map HdAccountId (Utxo,[HdAddress])
```
There are two changes here:

- The map is now indexed by `HdAccountBase` (which requires 'Ord' instance on this type, which is fine)
- The default account id and addresses are optional for E.O. wallets (more than optional, they actually don't exist), so, in order to save us some work here, I've simply remove them from the signature and it's now up to the caller to correctly construct the underlying Map (in most cases, a singleton with a default account id). 

This gives us an interoperable DB interface between both F.O. and E.O. wallets (cf higher layers below)

KtorZ commented on src/Cardano/Wallet/Kernel/DB/AcidState.hs:
```diff
@@ -615,20 +572,12 @@ restoreHdWallet newRoot defaultHdAccountId defaultHdAddress ctx utxoByAccount =
-          _accountUpdateBaseEOId             :: !HdAccountId
-        , _accountUpdateBaseEOPublicKey      :: !PublicKey
-        , _accountUpdateBaseEOAddressPoolGap :: !AddressPoolGap
-      }
```
This was actually not needed and doesn't buy us much since this is completely isomorphic to `HdAccountBase`.

KtorZ commented on src/Cardano/Wallet/Kernel/DB/AcidState.hs:
```diff
@@ -708,9 +653,6 @@ updateAccount AccountUpdate{..} = do
   where
-    getAccountId updateBase = case updateBase of
-        AccountUpdateBaseFO accId     -> accId
-        AccountUpdateBaseEO accId _ _ -> accId
```
As noted in a previous PR, a lens does a nicer job here instead of having to pattern match all the time :+1:

KtorZ commented on src/Cardano/Wallet/Kernel/DB/HdWallet.hs:
```diff
@@ -297,7 +298,8 @@ data HdAccountBase =
         , _hdAccountBaseEOAddressPoolGap :: !AddressPoolGap
-      }
+        }
+    deriving (Eq, Ord)
```
Needed to have `HdAccountBase` as a `Map` key.

KtorZ commented on src/Cardano/Wallet/Kernel/Wallets.hs:
```diff
@@ -462,7 +479,7 @@ updatePassword pw hdRootId oldPassword newPassword = do
                                        else HD.HasSpendingPassword lastUpdateNow
 
                                res <- update' (pw ^. wallets)
-                                              (UpdateHdRootPassword hdRootId hasSpendingPassword)
```
name conflicts with the lens `hdRootId` now imported :man_shrugging:

parsonsmatt commented on src/Cardano/Wallet/WalletLayer/Kernel/Wallets.hs:
```diff
@@ -152,31 +151,26 @@ createEosWallet :: MonadIO m
-            V1.neweoswalAddressPoolGap newEosWalletRequest
-        name = V1.neweoswalName newEosWalletRequest
-        assuranceLevel = V1.neweoswalAssuranceLevel newEosWalletRequest
+createEosWallet wallet req = runExceptT $ do
```
this looks significantly nicer :smile: :+1:

jonathanknowles committed:
Export the `DecodingError` type and data constructors.

jonathanknowles committed:
Rename function `parseDataPart` to `parseDataWithChecksumPart`.

The new name is more accurate. The previous name could be taken to imply
that the checksum is separate, which it is not.

jonathanknowles committed:
Check elementary preconditions earlier within function `Bech32.decode`.

This way, we can fail fast before performing any further parsing.

jonathanknowles committed:
Implement logic to detect the positions of errors within a Bech32 string.

This implementation is based upon the implementation found here:

https://github.com/sipa/bech32/blob/master/ecc/javascript/bech32_ecc.js

rvl commented on lib/bech32/src/Codec/Binary/Bech32/Internal.hs:
```diff
@@ -345,6 +377,241 @@ toBase256 :: [Word5] -> Maybe [Word8]
+                           Error Location Detection
+-------------------------------------------------------------------------------}
+
+gf_1024_exp :: Array Int Int
```
What are these tables? Are they something like `p^i` where `i` is the i'th entry in the table and `p` is some special number?

akegalj replied:
yes - it would be nice to have a comment/reference/description of these tables somewhere if there is a description available. I see the source is from https://github.com/sipa/bech32/blob/master/ecc/javascript/bech32_ecc.js but adding few words would help maintain it

jonathanknowles replied:
These tables are a direct translation of the tables found in the reference implementation here:

https://github.com/sipa/bech32/blob/master/ecc/javascript/bech32_ecc.js#L4
https://github.com/sipa/bech32/blob/master/ecc/javascript/bech32_ecc.js#L77

As for what they **do**: they seem to be a pair of lookup tables that store the values of a pair of pre-computed functions (an exponent and a logarithm). As to the exact nature of these functions, I currently do not know the answer. We would almost certainly have to ask the creator of the reference implementation. (https://github.com/sipa)

akegalj replied:
> We would almost certainly have to ask the creator of the reference implementation. 

I would go for this - if we can't find how tables are generated ourself. It is possible maintainer won't be responsive in which case we would try to figure it long term - but its worth a try to ping him

rvl commented on lib/bech32/src/Codec/Binary/Bech32/Internal.hs:
```diff
@@ -345,6 +377,241 @@ toBase256 :: [Word5] -> Maybe [Word8]
+    575, 992, 463, 983, 243, 360, 970, 350, 267, 615, 766, 494, 31, 1009,
+    452, 710, 552, 128, 612, 600, 275, 322, 193 ]
+
+syndrome :: (Bits a, Num a) => a -> a
```
What is this function?

It could also be shortened a little by using `foldr xor 0` and factoring the `if testBit residue n then s else 0`.

akegalj replied:
I see its a direct translation of https://github.com/sipa/bech32/blob/bdc264f84014c234e908d72026b7b780122be11f/ecc/javascript/bech32_ecc.js#L154 .

As this is translation from js -> haskell I would expect first iteration to have more js-ish style and I am ok with that. With more direct translation as here it is much easier to convince the reviewers that implementation is correct (by correct I mostly mean that it is a dirrect translation of reference implementation). On second or third implementation we can modify this to look more functional or more haskell-ish. 

Alternativelly we as reviewers might all want to fully understand underlying mechanisms of beach32 at this point and check it - but I think it is much more feasible to go with first approach at this stage.

jonathanknowles replied:
@rvl wrote:
> What is this function?

This is a direct translation of the corresponding function in the reference JavaScript implementation:
https://github.com/sipa/bech32/blob/master/ecc/javascript/bech32_ecc.js#L154

jonathanknowles replied:
> It could also be shortened a little by using `foldr xor 0` and factoring the `if testBit residue n then s else 0`.

That's a good idea. I originally wrote this in a style that was as close as possible to the original implementation (for easier verification). If I have time, I'll try to rewrite this in a shorter way.

rvl reviewed (commented):
Is nice and readable (except for `locateErrors`).

akegalj commented on lib/bech32/src/Codec/Binary/Bech32/Internal.hs:
```diff
@@ -98,21 +101,21 @@ mkHumanReadablePart hrp
         Right $ HumanReadablePart hrp
   where
-    (validPortion, invalidPortion) = BS.break (not . valid) hrp
+    invalidCharPositions = CharPosition . fst <$>
```
This looks like a good candidate for `NonEmpty` instead of current `[]` . Additionally this would simplify L104 (and maybe others) as you would be guaranteed to have failure there

jonathanknowles replied:
> This looks like a good candidate for `NonEmpty` instead of current `[]` . Additionally this would simplify L104 (and maybe others) as you would be guaranteed to have failure there

Actually, it's possible for the `locateErrors` function to return the empty list, if it is not able to reliably determine the locations of errors (even if the string is provably erroneous).

I'm going to add a comment to the `locateErrors` function to make this clear.

akegalj reviewed (commented):
I have just skimmed over it and it looks good.

To approve it I would have to go for another run to compare in more details with referance implementation (ping me when it is ready)

jonathanknowles committed:
Add comments clarifying the relationship of Haskell functions to their counterparts in the JavaScript reference implementation.

jonathanknowles committed:
Expose the Bech32 character set in a way that is more convenient for test code to consume.

jonathanknowles committed:
Add property tests to demonstrate that the Bech32 decoder correctly rejects corrupted strings.

jonathanknowles committed:
Add further documentation to functions within module `Bech32`.

jonathanknowles committed:
If the position of an erroneous character is detected, verify that it is detected at the correct location.

jonathanknowles commented on lib/bech32/src/Codec/Binary/Bech32/Internal.hs:
```diff
@@ -14,18 +14,19 @@
 --
--- [sipa/bech32](https://github.com/sipa/bech32/tree/bdc264f84014c234e908d72026b7b780122be11f/ref/haskell)
-
+-- [sipa/bech32](https://git.io/fj8FV)
```
These shortened URLs are an official part of GitHub (and are easier to paste into source code).

akegalj commented on lib/bech32/test/Codec/Binary/Bech32Spec.hs:
```diff
@@ -87,6 +96,116 @@ spec = do
+                Bech32.decode (getValidBech32String v) `shouldBe`
+                    Right (humanReadablePart v, unencodedDataPart v)
+
+    describe "Decoding a corrupted string should fail" $ do
```
did you have a look at these check bits https://github.com/sipa/bech32/blob/master/ecc/javascript/segwit_addr_ecc.js#L27 ? (maybe they are not relevant)

jonathanknowles replied:
@akegalj wrote:

> did you have a look at these check bits https://github.com/sipa/bech32/blob/master/ecc/javascript/segwit_addr_ecc.js#L27 ? (maybe they are not relevant)

The Haskell translation of that particular function is at the following location:
`lib/bech32/src/Codec/Binary/Bech32/Internal.hs:392`

The test for that function is here:
`lib/bech32/test/Codec/Binary/Bech32Spec.hs:235`

akegalj commented on lib/bech32/test/Codec/Binary/Bech32Spec.hs:
```diff
@@ -87,6 +96,116 @@ spec = do
             Bech32.encode hrp mempty
                 `shouldBe` Right (B8.pack "hrp1g9xj8m")
 
+    describe "Arbitrary ValidBech32String" $
```
would it make sense to ffi C implementation from https://github.com/sipa/bech32/blob/master/ref/c/segwit_addr.h and run property test comparing address encode/decode and encode/decode against our implementation ?

This way we would have complete test against implementation from another language. We could even use these implementation if there are more performant - but also we could only use these test in order to have complete run when we will be refactoring the implementation

jonathanknowles replied:
Actually, the most complete reference implementation seems to be the JavaScript one, which also provides error location detection. (The other implementations only provide a simple Boolean check to determine whether a given encoded string is valid.) I agree that a potentially useful side project would be to compare these two implementations. Perhaps it would be possible to use https://github.com/tweag/inline-js to make this easier.

akegalj reviewed (approved):
lgtm

I have partially looked and compared js and haskell implementation. As these bits are only relevant for error report and not actual encoding and decoding mechanism - it doesn't seem super important even if we did make an error in the translation js -> hs (which doesn't seems so from what I have seen)

Tests seem valuable

(there is a posibility to compare our bech32 encoder with other reference implementation out there, like C, if there would be suspicion that something is off. On another hand most of the implementation from https://github.com/sipa/bech32/ use same test vectors and I believe that should be enough)

jonathanknowles committed:
Unexport functions from Bech32Spec that are not called externally.
